File: 프로그래머스/lv2/42839. 소수 찾기/소수 찾기.py
import math
from itertools import permutations

def is_prime(number):
    if number < 2:
        return False
    if number == 2:
        return True
    
    for i in range(2, number):
        if number%i == 0:
            return False
    
    return True

def solution(numbers):
    answer = 0
    
    numbers = [number for number in numbers]
    num_list = set()
    
    for i in range(1, len(numbers) + 1):
        perm_list = list(permutations(numbers, i))
        for num in perm_list:
            num_list.add(int("".join(num)))
                         
    for num in num_list:
        print(num)
        if is_prime(num):
            answer += 1
        
    return answer

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    private static boolean[] visited;
    private static Set<Integer> set = new HashSet<>();
    
    public int solution(String numbers) {
        int answer = 0;
        
        visited = new boolean[numbers.length()];
        dfs(numbers, "");
        
        for (int number: set) {
            if (isPrime(number)) {
                answer++;
            }
        }
        
        return answer;
    }
    
    private static boolean isPrime(int number) {
        if (number < 2) {
            return false;
        }
        if (number == 2) {
            return true;
        }
        
        for (int n = 2; n <= Math.sqrt(number); n++) {
            if (number % n == 0) {
                return false;
            }
        }
        
        return true;
    }
    
    private static void dfs(String numbers, String number) {
        if (!number.isBlank()) {
            set.add(Integer.parseInt(number));
        }
                
        for (int i = 0; i < numbers.length(); i++) {
            if (!visited[i]) {
                visited[i] = true;
                dfs(numbers, number + numbers.charAt(i));
                visited[i] = false;
            }
        }
    }
}
/*
숫자가 적힌 문자열 numbers
만들 수 있는 소수 개수

*/